### DISCUSSION TITLE
Multi-objective search

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#736] [Q&A]
I have loss function, that contains 4 terms describing 4 different systems. While total loss seems fine, if one looks at each system individually the fit would be much worse in comparison to the result from only 2 terms ( if we compare two identical systems). 

It seems that multi objective optimization( pareto optimization ) is the solution.

Is it possible to use something like that? if not, will it be implemented in the near future?

**Comment by MilesCranmer** [comment]
Yes, the new `TemplateExpression` should help a lot with this kind of thing. It’s in the beta release now for SymbolicRegression. Unsure when or if I will make it available from PySR (since I think such flexibility would be incompatible with scikit-learn) but hopefully this will happen eventually.

**Reply by DenisSvirin** [reply]
Correct me if I'm wrong, but it seems that this helps users to search for a group of functions and evaluate these groups on aggregated loss (1 value).

I want loss function to return 4 losses and select better group based on my these 4 values + complexity, because it seems that combined loss doesn't work in my case.

**Reply by DenisSvirin** [reply]
It looks like i can modify `calculate_pareto_frontier` to select functions that decrease all 4 losses, but I'm afraid, that this will drastically cut possible solutions and algorithm wont work

**Reply by MilesCranmer** [reply]
> I want loss function to return 4 losses and select better group based on my these 4 values + complexity, because it seems that combined loss doesn't work in my case.

The one does this. You can define an elementwise loss function that works on vector output. 

Of course the actual minimum will be a sum of those 4 values + combined complexity. But I think that's the only thing possible?

**Reply by DenisSvirin** [reply]
Ah I see, Thank you!

I'm not quite familiar with the topic, but I've read about genetic algorithms, that use non dominant sorting (for example, NSGA-2) and i supposed that it can be used in pysr
